pkg/policy/api/rule_validation.go

```diff
@@ -113,84 +115,88 @@ func (i *IngressRule) sanitize() error {
 	for m1 := range l3Members {
 		for m2 := range l3Members {
 			if m2 != m1 && l3Members[m1] > 0 && l3Members[m2] > 0 {
-				return fmt.Errorf("Combining %s and %s is not supported yet", m1, m2)
+				return fmt.Errorf("Combining %s and %s is not supported yet", m1, m2), false
 			}
 		}
 	}
 
 	if len(l7Members) > 0 && !option.Config.EnableL7Proxy {
-		return errors.New("L7 policy is not supported since L7 proxy is not enabled")
+		return errors.New("L7 policy is not supported since L7 proxy is not enabled"), false
 	}
 	for member := range l7Members {
 		if l7Members[member] > 0 && !l7IngressSupport[member] {
-			return fmt.Errorf("L7 protocol %s is not supported on ingress yet", member)
+			return fmt.Errorf("L7 protocol %s is not supported on ingress yet", member), false
 		}
 	}
 
 	if len(i.ICMPs) > 0 && !option.Config.EnableICMPRules {
-		return fmt.Errorf("ICMP rules can only be applied when the %q flag is set", option.EnableICMPRules)
+		return fmt.Errorf("ICMP rules can only be applied when the %q flag is set", option.EnableICMPRules), false
 	}
 
 	if len(i.ICMPs) > 0 && len(i.ToPorts) > 0 {
-		return fmt.Errorf("The ICMPs block may only be present without ToPorts. Define a separate rule to use ToPorts.")
+		return fmt.Errorf("The ICMPs block may only be present without ToPorts. Define a separate rule to use ToPorts."), false
 	}
 
 	if len(i.FromNodes) > 0 && !option.Config.EnableNodeSelectorLabels {
-		return fmt.Errorf("FromNodes rules can only be applied when the %q flag is set", option.EnableNodeSelectorLabels)
+		if !softValidation {
+			return fmt.Errorf("FromNodes rules can only be applied when the %q flag is set", option.EnableNodeSelectorLabels), false
+		} else {
+			return fmt.Errorf("FromNodes rules can only be applied when the %q flag is set", option.EnableNodeSelectorLabels), true
```

Comment:
Hmm. The downside of this approach is that this will shortcut all of the remaining checks below.

Consider for the case where the user specifies this new flag on the CLI commandline, then loads a policy with two possible issues:
- The policy uses node selector labels, which may be OK if the server (cilium-agent) has a flag enabled
- The policy has an invalid `FromEndpoints` field.

In this scenario, the client will pass the flag and the client-side validation will exit early on this line, without checking `FromEndpoints`. Then the client will pass the same policy to the server, which will later fail on the `FromEndpoints` validation.

What do you think the behaviour should be in this case? Since the server still does the validation, I think that the behaviour will be correct (deny the invalid policy). In some ways it would be nicer if the client could validate the entire policy before sending it to the server.

Reply:
One option would be to declare this error statically at the top of the file:

```
errFromNodesRequiresNodeSelectorOption = fmt.Errorf("FromNodes rules can only be applied when the %q flag is set", option.EnableNodeSelectorLabels)
```

then,
```suggestion
			retErr = errFromNodesRequiresNodeSelectorOption
```

and continue to the end of the function. If any other error occurs, it would immediately return that "fatal" error. However, if the control reaches the end of the function, it could just return this `retErr` to the caller.

From the caller, it could check `if errors.Is(err, api.errFromNodesRequiresNodeSelectorOption) { // Don't error out }`.

In this case, we might not even need the new skipped boolean return value.

Reply:
yeah, this is so much better this way :+1: